src/lib/operators/join_sort_merge.cpp

```diff
@@ -497,26 +494,47 @@ class JoinSortMerge::JoinSortMergeImpl : public AbstractReadOnlyOperatorImpl {
         _sorted_right_table, [&](RowID right_row_id) { _emit_combination(output_cluster, NULL_ROW_ID, right_row_id); });
   }
 
-  /**
-  * Determines the length of the run starting at start_index in the values vector.
-  * A run is a series of the same value.
-  **/
-  size_t _run_length(size_t start_index, std::shared_ptr<MaterializedSegment<T>> values) {
-    if (start_index >= values->size()) {
+  // Determines the length of the run starting at start_index in the values vector. A run is a series of the same
+  // value. Even though the input vector is sorted, we first start by linearly scanning for the end of the run. The
+  // reason is that we know with a high probability that the item we are looking for is at the beginning of the input
+  // vector. If we do not find the run's end within the linearly scanned part, we use a binary search.
+  size_t _run_length(size_t start_index, const MaterializedSegment<T>& values) {
+    if (start_index >= values.size()) {
       return 0;
     }
 
-    auto start_position = values->begin() + start_index;
-    auto result = std::upper_bound(start_position, values->end(), *start_position,
-                                   [](const auto& a, const auto& b) { return a.value < b.value; });
+    const auto begin = values.begin() + start_index;
+    const auto run_value = begin->value;
+
+    constexpr auto LINEAR_SEARCH_ITEMS = size_t{128};
```

Comment:
Are you sure that this is more performant than just a binary search from the beginning? With 128 comparisons you can sift through a huge amount of values with binary search.

Reply:
Yep, measured. The problem with binary searches is that you basically have a cache miss for every check. In this case here, we know that the searched value is with a high probability in the beginning. If it's  not, the cache is in the cache and we are going to iterate over it anyways. To the costs for a non-match are neglectable.

Reply:
Does the added comment `// An unsuccessful linear search should come with neglectable costs as we will iterate over the values anyways.` help?

Reply:
Okay, if that is true, we should scan the 128 values for every step of the binary search, since we have a scan 1 and get 128 for free situation. Does std::distance to that or does it only look at 1 value at each step of the binary search?

Reply:
Some binary searches actually do that and access aligned cache line-sized items and check them with SIMD. But I am pretty sure that code would not make into `std::` and it does not make sense for us to implement our own binary search.
More importantly: we get the items "for free" because we scan them afterwards anyways. That's a different case than a typical binary search to find an item.

We once measured the performance of binary search vs. linear search and if I remember it correctly binary search won when the vector contained a couple of hundred items.